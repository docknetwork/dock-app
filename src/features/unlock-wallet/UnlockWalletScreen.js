import React, {useCallback, useEffect, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {NumericKeyboard} from 'src/components/NumericKeyboard';
import {showToast} from 'src/core/toast';
import styled from 'styled-components/native';
import SplashLogo from '../../assets/splash-logo.png';
import {
  Box,
  Content,
  Image,
  ScreenContainer,
  Typography,
} from '../../design-system';
import {translate} from '../../locales';
import {appSelectors} from '../app/app-slice';
import {BuildIdentifier} from '../app/BuildIdentifier';
// import {NumericKeyboard} from '../wallet/CreatePasscodeScreen';
import {walletOperations, walletSelectors} from '../wallet/wallet-slice';

const Circle = styled.View`
  width: 20px;
  height: 20px;
  border-radius: 20px;
  border: 1px solid white;
  margin: 0 7px;
  background-color: ${props => (props.filled ? '#fff' : 'transparent')};
`;

function PasscodeMask({digits = 6, filled = 0, ...props}) {
  const digitsArray = new Array(digits).fill(0);

  return (
    <Box {...props}>
      <Box justifyContent="center" row>
        {digitsArray.map((_, idx) => (
          <Circle filled={idx < filled} key={idx} />
        ))}
      </Box>
    </Box>
  );
}

const DIGITS = 6;

export function UnlockWalletScreen({
  onPasscodeChange,
  passcode,
  onLoginWithBiometric,
  text = translate('unlock_wallet.enter_passcode'),
  digits = DIGITS,
  filled = 0,
  biometry = false,
}) {
  return (
    <ScreenContainer testID="unlockWalletScreen">
      <Box justifyContent="center" row>
        <Image
          source={SplashLogo}
          style={{
            width: '57%',
          }}
          resizeMode="contain"
        />
      </Box>
      <Content marginLeft={26} marginRight={26}>
        <PasscodeMask marginTop={20} digits={digits} filled={filled} />
        <Box alignItems="center">
          <Typography variant="h2" marginTop={52}>
            {text}
          </Typography>
        </Box>
        <NumericKeyboard
          marginTop={50}
          onChange={onPasscodeChange}
          value={passcode}
        />
        {biometry ? (
          <Box
            alignItems="center"
            marginBottom={20}
            onPress={onLoginWithBiometric}>
            <Typography variant="description">
              {translate('unlock_wallet.biometrics_login')}
            </Typography>
          </Box>
        ) : null}
        <BuildIdentifier />
      </Content>
    </ScreenContainer>
  );
}

export function UnlockWalletContainer({route}) {
  const {callback} = route.params || {};
  const [passcode, setPasscode] = useState('');
  const supportBiometry = useSelector(appSelectors.getSupportedBiometryType);
  const walletInfo = useSelector(walletSelectors.getWalletInfo);
  const biometryEnabled = walletInfo && walletInfo.biometry;

  const dispatch = useDispatch();

  const handlePasscodeChange = async value => {
    if (value.length > DIGITS) {
      return;
    }

    setPasscode(value);

    if (value.length === DIGITS) {
      try {
        await dispatch(
          walletOperations.unlockWallet({passcode: value, callback}),
        );
      } catch (err) {
        showToast({
          message: translate('unlock_wallet.invalid_passcode'),
          type: 'error',
        });
      }

      setTimeout(() => {
        setPasscode('');
      }, 100);
    }
  };

  const handleBiometricUnlock = useCallback(async () => {
    try {
      await dispatch(walletOperations.unlockWallet({biometry: true, callback}));
    } catch (err) {
      setPasscode('');
    }
  }, [dispatch, callback]);

  useEffect(() => {
    if (supportBiometry && biometryEnabled) {
      handleBiometricUnlock();
    } else {
      setPasscode('');
    }
  }, [supportBiometry, biometryEnabled, handleBiometricUnlock]);

  return (
    <UnlockWalletScreen
      digits={DIGITS}
      filled={passcode.length}
      onPasscodeChange={handlePasscodeChange}
      onLoginWithBiometric={handleBiometricUnlock}
      biometry={supportBiometry}
      passcode={passcode}
    />
  );
}
