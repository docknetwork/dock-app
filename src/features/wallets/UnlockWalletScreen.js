import {
  Picker,
  Header,
  Footer,
  Content,
  Button,
  Text,
  Container,
  Spinner,
  View,
  Form,
  Item,
  Label,
  Input,
  Grid,
  Row,
  Col,
  H2,
  Icon,
} from 'native-base';
import React, {useEffect, useState} from 'react';
import {Dimensions, InteractionManager} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {ScreenSpinner} from '../../components/ScreenSpinner';
import {navigate} from '../../core/navigation';
import {Routes} from '../../core/routes';
import {Colors} from '../../theme/colors';
import {walletsOperations, walletsSelectors} from './wallets-slice';

export function WalletPicker({
  address,
  password,
  onAddressChange,
  onPasswordChange,
  error,
}) {
  const wallets = useSelector(walletsSelectors.getItems);

  return (
    <Form>
      <Item picker>
        <Picker
          note
          mode="dropdown"
          style={{width: Dimensions.get('screen').width, height: 70}}
          iosIcon={<Icon name="arrow-down" />}
          placeholder="Select a wallet"
          placeholderStyle={{color: '#333'}}
          selectedValue={address}
          onValueChange={onAddressChange}>
          {wallets.map(wallet => (
            <Picker.Item
              label={
                <View style={{paddingTop: 8, paddingLeft: 12}}>
                  <Text style={{fontWeight: 'bold'}}>{wallet.meta.name}</Text>
                  <Text style={{color: '#555', fontSize: 12}}>
                    {wallet.address}
                  </Text>
                </View>
              }
              value={wallet.address}
              key={wallet.address}
            />
          ))}
        </Picker>
      </Item>
      {address ? (
        <Item floatingLabel>
          <Label>Enter the password</Label>
          <Input
            onChangeText={onPasswordChange}
            value={password}
            secureTextEntry={true}
            error={error}
          />
        </Item>
      ) : null}
    </Form>
  );
}

export function UnlockWalletScreen() {
  const dispatch = useDispatch();
  const [selectedAddress, setSelectedAddress] = useState();
  const wallets = useSelector(walletsSelectors.getItems);
  const [password, setPassword] = useState();
  const [error, setError] = useState();
  const [loading, setLoading] = useState();
  const handlePasswordChange = value => setPassword(value);

  const handleCreateWallet = () => {
    navigate(Routes.CREATE_WALLET);
  };
  const handleUnlockWallet = async () => {
    setError(null);
    setLoading(true);

    InteractionManager.runAfterInteractions(() => {
      dispatch(
        walletsOperations.unlockWallet({address: selectedAddress, password}),
      )
        .then(() => {
          setPassword(null);
          setLoading(false);
        })
        .catch(() => {
          setLoading(false);
          setError('Invalid password, please try again.');
        });
    });
  };

  useEffect(() => {
    if (!wallets) {
      return;
    }

    if (!wallets.length) {
      navigate(Routes.CREATE_WALLET);

      return;
    }

    if (!selectedAddress) {
      setSelectedAddress(wallets[0].address);
    }
  }, [wallets]);

  if (!wallets || !wallets.length) {
    return (
      <Container>
        <Content>
          <ScreenSpinner />
        </Content>
      </Container>
    );
  }

  return (
    <Container>
      <Header style={{alignItems: 'center', marginTop: 10}}>
        <H2 style={{color: '#fff'}}>Select a Wallet</H2>
      </Header>
      <Content>
        <WalletPicker
          password={password}
          address={selectedAddress}
          onPasswordChange={handlePasswordChange}
          onAddressChange={v => setSelectedAddress(v)}
          error={error}
        />
        <View style={{padding: 20}}>
          {<Text style={{color: Colors.red}}>{error}</Text>}
        </View>
      </Content>
      <Footer style={{backgroundColor: '#fff', padding: 8, height: 120}}>
        <Grid>
          <Row>
            <Col>
              <Button
                full
                style={{width: '100%'}}
                primary
                onPress={handleUnlockWallet}
                disabled={!selectedAddress || loading}>
                {loading ? (
                  <Spinner color="#fff" size={15} />
                ) : (
                  <Text style={{color: '#fff'}}>Unlock Wallet</Text>
                )}
              </Button>
            </Col>
          </Row>
          <Row>
            <Col>
              <Button
                full
                style={{width: '100%'}}
                bordered
                onPress={handleCreateWallet}>
                <Text style={{color: Colors.darkBlue}}>Create new wallet</Text>
              </Button>
            </Col>
          </Row>
        </Grid>
      </Footer>
    </Container>
  );
}
