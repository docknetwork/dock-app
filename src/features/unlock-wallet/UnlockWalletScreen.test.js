import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {UnlockWalletScreen} from './UnlockWalletScreen';
import {
  authenticationReducer,
  authenticationActions,
} from './unlock-wallet-slice';

const mockStore = configureMockStore();

describe('UnlockWalletScreen', () => {
  it('should render correctly', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(<UnlockWalletScreen biometry={true} />, {
      context: {store: mockStore(initialState)},
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
});

describe('Test Authentication', function () {
  it('should return the initial state', () => {
    expect(authenticationReducer(undefined, {})).toEqual({
      isLoggedIn: false,
    });
  });

  it('should handle auth state change', () => {
    const previousState = {
      isLoggedIn: false,
    };
    expect(
      authenticationReducer(
        previousState,
        authenticationActions.setAuth({isLoggedIn: true}),
      ),
    ).toEqual({
      isLoggedIn: true,
    });
  });
});
