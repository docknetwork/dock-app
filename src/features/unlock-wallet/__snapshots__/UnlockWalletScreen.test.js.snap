// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnlockWalletScreen should render correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#18181B",
      "flex": 1,
    }
  }
  testID="unlockWalletScreen"
>
  <Box
    justifyContent="center"
    row={true}
  >
    <Styled(Image)
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/assets/splash-logo.png",
        }
      }
      style={
        Object {
          "width": "57%",
        }
      }
    />
  </Box>
  <Styled(ScrollView)
    marginLeft={26}
    marginRight={26}
  >
    <PasscodeMask
      digits={6}
      filled={0}
      marginTop={20}
    />
    <Box
      alignItems="center"
    >
      <Styled(Component)
        marginTop={52}
        variant="h2"
      >
        Enter your passcode
      </Styled(Component)>
    </Box>
    <NumericKeyboard
      marginTop={50}
    />
    <Box
      alignItems="center"
      marginBottom={20}
    >
      <Styled(Component)
        variant="description"
      >
        Login with biometrics
      </Styled(Component)>
    </Box>
    <BuildIdentifier />
    <Modal>
      <Memo(ForwardRef(Stack))
        alignItems="center"
        m={5}
        p={5}
      >
        <Styled(Component)
          variant="h2"
        >
          App locked
        </Styled(Component)>
        <Styled(Component)
          variant="h4"
        >
          Please try again later
        </Styled(Component)>
        <Memo(ForwardRef(Stack))
          mt={4}
        >
          <Button>
            Close app
          </Button>
        </Memo(ForwardRef(Stack))>
      </Memo(ForwardRef(Stack))>
    </Modal>
  </Styled(ScrollView)>
</RNCSafeAreaView>
`;
